net: if: set NET_ADDR_PREFERRED before network event if !IPV4_ACD and/or !IPV6_DAD

mniestroj opened the pull request:
Set `NET_ADDR_PREFERRED` before network event is generated, so that conn_mgr properly generates
`NET_EVENT_L4_CONNECTED` event.

Fixes: 1a5e13a79b98 ("net: if: Release the interface lock early when
  starting IPv4 ACD")

Tested with and with following snippet:
```
	if (mgmt_event == NET_EVENT_L4_IPV4_CONNECTED) {
		LOG_INF("IPv4 connected");
	}

	if (mgmt_event == NET_EVENT_L4_IPV4_DISCONNECTED) {
		LOG_INF("IPv4 disconnected");
	}

	if (mgmt_event == NET_EVENT_L4_IPV6_CONNECTED) {
		LOG_INF("IPv6 connected");
	}

	if (mgmt_event == NET_EVENT_L4_IPV6_DISCONNECTED) {
		LOG_INF("IPv6 disconnected");
	}
```
to have more events generated. Used `west build -p -b native_sim/native/64 zephyr/samples/net/sockets/echo_server -- -DEXTRA_CONF_FILE=overlay-nsos.conf` for testing.

Before this patch:
```
[00:00:00.000,000] <inf> net_config: Initializing network
[00:00:00.000,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:00.000,000] <inf> net_echo_server_sample: Run echo server
[00:00:00.000,000] <inf> net_echo_server_sample: Waiting network to be connected
[00:00:00.000,000] <inf> net_echo_server_sample: IPv6 disconnected
[00:00:00.000,000] <inf> net_echo_server_sample: IPv4 disconnected
```
after those patches:
```
[00:00:00.000,000] <inf> net_config: Initializing network
[00:00:00.000,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:00.000,000] <inf> net_echo_server_sample: Run echo server
[00:00:00.000,000] <inf> net_echo_server_sample: Network connected
[00:00:00.000,000] <inf> net_echo_server_sample: IPv6 connected
[00:00:00.000,000] <inf> net_echo_server_sample: IPv4 connected
[00:00:00.000,000] <inf> net_echo_server_sample: Starting...
[00:00:00.000,000] <inf> net_echo_server_sample: Waiting for TCP connection on port 4242 (IPv6)...
[00:00:00.000,000] <inf> net_echo_server_sample: Waiting for TCP connection on port 4242 (IPv4)...
[00:00:00.000,000] <inf> net_echo_server_sample: Waiting for UDP packets on port 4242 (IPv6)...
[00:00:00.000,000] <inf> net_echo_server_sample: Waiting for UDP packets on port 4242 (IPv4)...
```

Fixes: #87223

mniestroj committed:
net: ipv4: set NET_ADDR_PREFERRED before network event if !ACD

Set NET_ADDR_PREFERRED before network event is generated, so that conn_mgr
properly generates NET_EVENT_L4_CONNECTED and NET_EVENT_L4_IPV4_CONNECTED
events.

Fixes: 1a5e13a79b98 ("net: if: Release the interface lock early when
  starting IPv4 ACD")
Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

rlubos commented:
Can you open an issue for this, it seems that we'll need to backport the fix.

It also seems that IPv6 would have the same problem, any chance to get that addressed too?

mniestroj committed:
net: ipv6: set NET_ADDR_PREFERRED before network event if !DAD

Set NET_ADDR_PREFERRED before network event is generated, so that conn_mgr
properly generates NET_EVENT_L4_CONNECTED and NET_EVENT_L4_IPV6_CONNECTED
events.

Fixes: 1a5e13a79b98 ("net: if: Release the interface lock early when
  starting IPv4 ACD")
Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
samples: net: echo_server: add runtime test for NSOS net init

Expect that networking is properly setup, by expecting
NET_EVENT_L4_CONNECTED event and listening for TCP/UDP incoming traffic on
both IPv4 and IPv6.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj commented:
Added #86870 as requested.

mniestroj commented:
Added #87223 as requested

rlubos commented:
> Added #87223 as requested

Thanks!
